aim/web/ui/src/modules/BaseExplorer/components/BoxVirtualizer/BoxVirtualizer.tsx

```diff
@@ -85,28 +99,81 @@ function BoxVirtualizer(props: IBoxVirtualizerProps) {
       ),
     [filteredItems],
   );
-  // @TODO remove this variable and calculate width and height of the container with optimized way
-  const sortedByPosition = React.useMemo(
-    () =>
-      data
-        .sort((a: any, b: any) => a.style.left - b.style.left)
-        .sort((a: any, b: any) => a.style.top - b.style.top),
-    [data],
-  );
-  const gridWidth =
-    sortedByPosition?.[sortedByPosition?.length - 1]?.style?.left +
-    sortedByPosition?.[sortedByPosition?.length - 1]?.style?.width;
-  const gridHeight =
-    sortedByPosition?.[sortedByPosition?.length - 1]?.style?.top +
-    sortedByPosition?.[sortedByPosition?.length - 1]?.style?.height;
+
+  // Find the edges for container size calculation
+  const gridSize = React.useMemo(() => {
+    let rightEdge = 0;
+    let bottomEdge = 0;
+
+    let itemWidth = 0;
+    let itemHeight = 0;
+
+    for (let i = 0; i < data.length; i++) {
+      let item = data[i];
+      if (item.style.top > rightEdge) {
+        rightEdge = item.style.left;
+        itemWidth = item.style.width;
+      }
+
+      if (item.style.top > bottomEdge) {
+        bottomEdge = item.style.top;
+        itemHeight = item.style.height;
+      }
+    }
+
+    const horizontalRulerHeight = 30;
+
+    return {
+      width: rightEdge + itemWidth + props.offset,
+      height: bottomEdge + itemHeight + props.offset - horizontalRulerHeight,
+    };
+  }, [data, props.offset]);
+
   return (
-    <div ref={container} className='BoxVirtualizer' onScroll={onScroll}>
+    <div className='BoxVirtualizer'>
+      {columnsAxisItems &&
+        columnsAxisItems.length > 0 &&
+        rowsAxisItems &&
+        rowsAxisItems.length > 0 && (
+          <div className='BoxVirtualizer__placeholder' />
+        )}
       <div
-        ref={grid}
-        className='BoxVirtualizer__grid'
-        style={{ width: gridWidth || 0, height: gridHeight || 0 }}
+        ref={container}
+        className='BoxVirtualizer__container'
+        onScroll={onScroll}
       >
-        {Object.entries(groupedByPosition).map(props.itemsRenderer)}
+        {((columnsAxisItems && columnsAxisItems.length > 0) ||
+          (rowsAxisItems && rowsAxisItems.length > 0)) && (
+          <div
+            className='BoxVirtualizer__container__horizontalRuler'
+            style={{
+              width: gridSize.width,
+            }}
+          >
+            {columnsAxisItems?.map(props.axisItemRenderer?.columns)}
+          </div>
+        )}
+        {rowsAxisItems && rowsAxisItems.length > 0 && (
+          <div
+            className='BoxVirtualizer__container__verticalRuler'
+            style={{
+              height: gridSize.height,
+            }}
+          >
+            {rowsAxisItems?.map(props.axisItemRenderer?.rows)}
+          </div>
+        )}
+        <div
+          ref={grid}
+          style={{
+            display: 'inline',
```

Comment:
let's add the class name here and include static styles in it,  remember that I have added the one named `grid` which we remove maybe after resolving the conflicts